text/0000-builder-naming.md

```diff
@@ -0,0 +1,136 @@
+
+   but *not* `fs::Builder`, which opens files;
+* If the builder produces one type of objects, use the name of buildee suffixed with “Builder”.
+    This scheme has an advantage of being searchable and self-documenting. For example:
+    `FileBuilder` is a builder and produces objects of type `File`;
+* If there is no clear buildee (most likely because it contains multiple finalizers), a noun
+    describing all the buildable objects suffixed with “Builder” may be used. For example:
+    `ThreadBuilder`, even though it produces `JoinHandle`s and `JoinGuard`s rather than `Thread`s;
+* Finally, if the majority (community) decides there exists a superior name for the builder, but it
+    does not match any of the guidelines above, generously point out the existence of the builder
+    in the documentation of buildee and module, if appropriate.
+
+### Methods
+
+Option setters – methods which represent options that can be set on the builder – should receive
+the builder object as `&mut self` rather than `self` unless there is a good reason to do otherwise.
+
+Finalizers – methods which produce the built object – should receive self via reference, unless
+there is a good reason to do otherwise.
+
+In order for method listing in documentation to be predictable, methods in the builder’s `impl`
+block should be declared following this order:
+
+* `new` – a method to create the builder object comes first in the `impl` block;
+* followed by one or more option setters;
+* followed by one or more finalizers.
+
+This ordering stems from conventional use of the builder pattern: create a builder, set the
+options and call a finalizer.
+
+#### Method naming
+
+Option setters should be nouns or adjectives. Avoid `is_*` prefix and nouns derived from
+adjectives. For example:
+
+* `name`;
+* `stack_size`;
+* `read` or `readable` (rather than `is_readable` or `readability`);
+* `write` or `writable` (rather than `is_writable` or `writability`).
+
+Finalizers should be verbs. For example:
+
+* `spawn`;
+* `create`;
+* `open`.
+
+## Proposed changes to the standard library
+
+### `thread::Builder`
+
+`thread::Builder` is adjusted to have the following interface:
+
+```rust
+pub struct Builder { … }
+impl Builder {
+    fn new() → Builder;
+    fn name(&mut self, name: String) → &mut Builder;
+    fn stack_size(&mut self, size: usize) → &mut Builder;
+    fn spawn<F>(&mut self, f: F) → Result<JoinHandle> where F: FnOnce(), F: Send + 'static;
+    fn scoped<'a, T, F>(&mut self, f: F) → Result<JoinGuard<'a, T>>
+        where T: Send + 'a, F: FnOnce() → T, F: Send + 'a
+}
+```
+
+and `impl Clone for Builder` is provided.
+
+In summary, all methods are changed to take and return `Builder` by mutable reference rather than
+by value.  Strictly speaking this is a breaking change, but most users of the `thread::Builder`
+should not encounter any breakage.
+
+### `process::Command`
+
+`impl Clone for Command` is provided.
```

Comment:
[In the past](https://github.com/rust-lang/rust/pull/22858) I've been hesitant to do this as it may prevent us from adding more options to `Command` in the future. For example if a standard I/O handle is [redirected to a file](https://github.com/rust-lang/rfcs/pull/1055), it's not clear how `Clone` can be implemented.

We do have various routes to mitigate this, but this may fall under the category of "would like clone, but implementation may not make it possible".